File: tree_zig_zag/solutions/tree_zig_zag.py
#!/usr/bin/env python

import collections
import sys


class Node(object):
    __slots__ = ('value', 'left', 'right')

    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

    def __hash__(self):
        return hash(self.value)

    def __eq__(self, other):
        if id(self) == id(other):
            return True
        return self.value == other.value

    def __str__(self):
        return "{value=%s, left=%s, right=%s}" % (self.value, self.left, self.right)

    def __repr__(self):
        return str(self.value)


class TreeReader(object):
    @classmethod
    def get_vertex(cls, value, vertices):
        if value == -1:
            return None
        key = Node(value)
        if key not in vertices:
            vertices[key] = key
        return vertices[key]

    @classmethod
    def read_tree(cls, stream):
        root = None
        vertices = {}
        for line in stream:
            (vertex_value, left_value, right_value) = map(int, line.split())
            vertex = cls.get_vertex(vertex_value, vertices)
            vertex.left = cls.get_vertex(left_value, vertices)
            vertex.right = cls.get_vertex(right_value, vertices)
            if root is None:
                root = vertex
        return root


def zig_zag_traverse(root):
    is_left_most = True
    current_level = []
    sentinel = Node(-1)
    queue = collections.deque([root, sentinel])
    while len(queue) != 0:
        node = queue.popleft()
        if node == sentinel:
            index = 0 if is_left_most else len(current_level) - 1
            print(current_level[index].value)
            current_level = []
            is_left_most = not is_left_most
            if len(queue) != 0:
                queue.append(sentinel)
        else:
            current_level.append(node)
            queue.extend(x for x in [node.left, node.right] if x)


def main():
    root = TreeReader.read_tree(sys.stdin)
    zig_zag_traverse(root)


if __name__ == "__main__":
    sys.exit(main())

File: tree_zig_zag/solutions/TreeZigZag.java
import java.io.InputStream;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Queue;
import java.util.Scanner;


class Node<T> {
    private T value;
    private Node<T> left;
    private Node<T> right;

    /**
     * An empty constructor is used only for creating sentinels, which
     * are useful in some algorithms. This is why a corresponding
     * setValue() is not provided.
     */
    Node() {
    }

    Node(T value) {
        this.value = value;
    }

    public void setLeft(Node<T> left) { this.left = left; }
    public Node<T> getLeft() { return this.left; }
    public void setRight(Node<T> right) { this.right = right; }
    public Node<T> getRight() { return this.right; }
    public T getValue() { return this.value; }

    @Override
    public String toString() {
        return String.format("{value=%s, left=%s, right=%s}", value, left, right);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Node<?>)) return false;
        Node<?> node = (Node<?>)o;
        return this.value.equals(node.value);
    }

    @Override
    public int hashCode() {
        int result = 17;
        result = result * 31 + value.hashCode();
        return result;
    }
}


class ZigZagTreeLevelSearch<T> {
    private boolean isLeftMost;

    public void search(Node<T> root) {
        List<Node<T>> currentLevel = new ArrayList<Node<T>>();
        isLeftMost = true;
        Node<T> sentinel = new Node<T>();
        Queue<Node<T>> queue = new ArrayDeque<Node<T>>();
        queue.add(root);
        queue.add(sentinel);
        while(!queue.isEmpty()) {
            Node<T> current = queue.remove();
            if (current == sentinel) {
                processLevel(currentLevel);
                isLeftMost = !isLeftMost;
                currentLevel = new ArrayList<Node<T>>();
                if (!queue.isEmpty())
                    queue.add(sentinel);
            } else {
                currentLevel.add(current);
                if (current.getLeft() != null)
                    queue.add(current.getLeft());
                if (current.getRight() != null)
                    queue.add(current.getRight());
            }
        }
    }

    public void processLevel(List<Node<T>> level) {
        int index = isLeftMost ? 0 : level.size() - 1;
        System.out.println(level.get(index).getValue());
    } 
}


class TreeReader {
    public Node<Integer> getOrCreate(Integer value, Map<Node<Integer>, Node<Integer>> map) {
        if (value.equals(-1))
            return null;
        Node<Integer> key = new Node<Integer>(value);
        Node<Integer> node;
        if (map.containsKey(key)) {
            node = map.get(key);
        } else {
            node = key;
            map.put(key, key);
        }
        return node;
    }

    public Node<Integer> readTree(InputStream in) {
        Map<Node<Integer>, Node<Integer>> map = new HashMap<Node<Integer>, Node<Integer>>();
        Node<Integer> root = null;
        Scanner sc = new Scanner(in);
        while (sc.hasNextLine()) {
            Integer vertexValue = sc.nextInt();
            Integer leftValue = sc.nextInt();
            Integer rightValue = sc.nextInt();
            Node<Integer> vertex = getOrCreate(vertexValue, map);
            Node<Integer> left = getOrCreate(leftValue, map);
            Node<Integer> right = getOrCreate(rightValue, map);
            vertex.setLeft(left);
            vertex.setRight(right);
            if (root == null)
                root = vertex;
        }
        return root;
    }    
}


class TreeZigZag {
    public static void main(String args[]) {
        TreeReader tr = new TreeReader();
        Node<Integer> root = tr.readTree(System.in);
        ZigZagTreeLevelSearch<Integer> searcher = new ZigZagTreeLevelSearch<Integer>();
        searcher.search(root);
    }
}
